internal/cluster-object/secret_provider_test.go

```diff
@@ -0,0 +1,315 @@
+		siProvider := NewServiceInstanceProvider(k8sClient, logger)
+		secretProvider := NewSecretProvider(k8sClient, nsProvider, siProvider, logger)
+
+		// when
+		actualSecrets, err := secretProvider.All(context.TODO())
+		if err != nil {
+			t.Errorf("Error while fetching secrets: %s", err)
+		}
+
+		// then
+		compareSecretSlices(t, expectedSecrets, actualSecrets.Items)
+	})
+
+	t.Run("should return nil when there are no btp operator secrets", func(t *testing.T) {
+		// given
+		additionalNamespaces := createAdditionalNamespaces()
+		additionalSecrets := createAdditionalSecrets()
+		secrets := &corev1.SecretList{Items: additionalSecrets}
+		ns.Items = append(ns.Items, additionalNamespaces...)
+
+		k8sClient := fake.NewClientBuilder().
+			WithLists(ns, sis, secrets).
+			WithIndex(&corev1.Secret{}, "metadata.name", secretNameIndexer).
+			Build()
+		nsProvider := NewNamespaceProvider(k8sClient, logger)
+		siProvider := NewServiceInstanceProvider(k8sClient, logger)
+		secretProvider := NewSecretProvider(k8sClient, nsProvider, siProvider, logger)
+
+		// when
+		actualSecrets, err := secretProvider.All(context.TODO())
+		if err != nil {
+			t.Errorf("Error while fetching secrets: %s", err)
+		}
+
+		// then
+		if actualSecrets != nil {
+			t.Errorf("Expected nil, got %T", actualSecrets)
+		}
+	})
+}
+
+func createAdditionalNamespaces() []corev1.Namespace {
+	return []corev1.Namespace{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: controllers.ChartNamespace,
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "test1",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "test2",
+			},
+		},
+	}
+}
+
+func createAdditionalSecrets() []corev1.Secret {
+	return []corev1.Secret{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "unrelated-additional-secret1",
+				Namespace: controllers.ChartNamespace,
+			},
+			StringData: map[string]string{
+				"foo": "bar",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "unrelated-additional-secret2",
+				Namespace: "test1",
+			},
+			StringData: map[string]string{
+				"foo": "bar",
+			},
+		},
+	}
+}
+
+func secretNameIndexer(obj client.Object) []string {
+	secret, ok := obj.(*corev1.Secret)
+	if !ok {
+		panic(fmt.Errorf("indexer function for type %T's metadata.name field received"+
+			" object of type %T, this should never happen", corev1.Secret{}, obj))
+	}
+
+	return []string{secret.Name}
+}
+
+func compareSecretSlices(t *testing.T, expected, actual []corev1.Secret) {
+	if len(expected) != len(actual) {
+		t.Errorf("Expected %d secrets, got %d", len(expected), len(actual))
+		return
+	}
+
+	for _, expectedSecret := range expected {
+		if !containsSecret(actual, expectedSecret) {
+			t.Errorf("Expected secret %s not found in the actual list", expectedSecret.Name)
+		}
+	}
+}
+
+func containsSecret(secrets []corev1.Secret, secret corev1.Secret) bool {
+	for _, s := range secrets {
+		if s.Name == secret.Name && s.Namespace == secret.Namespace {
```

Comment:
I don't think so, we need to check every element in the slice. Inline check with return will check just the first element. I could wrap this fragment in a lambda, but this will increase the complexity.